### PR TITLE
Test timer

### PR DESCRIPTION
**Pull request by skothari-tibco**
Add trigger similar to CSV Timer. It takes in csv file and outputs each line of file at a certain interval. Control API is also added.

### CODE REVIEW

File: trigger/test-timer/README.md [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+  }
+}
+```
+### Details
+
+#### Settings:
+| Setting  | Required | Description |
+|:---------|:---------|:------------|
+| control  | false    | Enable control of trigger
+| port     | false    | Port number on which control api is called (/control/resume, /control/pause, /control/restart)
```
[the remark this answers is not in the document]
Comment (fm-tibco) [reply trigger/test-timer/README.md]: We can add an autostart flag or something.  But definitely start/stop.  The pause/resume only make sense if you are parsing the csv file and emmitting a value every (1ms, 10ms).  So I think we should have a mode where you are parsing the csv file and emitting rows.
So you can have an emit data time interval, (1ms.. 10ms.. etc.) and a flag or enum to indicate whether to emit rows or the entire file.    

In this case of rows, pause/resume would make more sense, because otherwise they are essentially just start and stop.

[the remark this answers is not in the document]
Comment (fm-tibco) [reply trigger/test-timer/README.md]: I would remove the timer stuff, this is essentially a loop for each handler with a sleep in it.  Also I would add the option of specifying a handler id when invoking the control rest commands.  That way if someone wants to pause a particular handler they can.

[the remark this answers is not in the document]
Comment (skothari-tibco) [reply trigger/test-timer/README.md]: If we remove the timer stuff, how will we decide, the interval after which the data should be given as output?

[the remark this answers is not in the document]
Comment (fm-tibco) [reply trigger/test-timer/README.md]: simple loop with a sleep would probably suffice, you just sleep for that interval

File: go.mod [hunk 2]
```diff
@@ -1,7 +1,9 @@
 module github.com/project-flogo/stream
 
 require (
```
Comment (fm-tibco) [inline go.mod]: Revert this changes, stream doesn't use these.

File: pipeline/support/mapper.go [hunk 3]
```diff
@@ -70,10 +70,10 @@ func (m *OverrideExprMapper) Apply(inputScope data.Scope) (map[string]interface{
 	for key, expr := range m.mappings {
 		val, err := expr.Eval(inputScope)
 		if err != nil {
+
```
Comment (fm-tibco) [inline pipeline/support/mapper.go]: exclude this file from this PR

File: trigger/testTimer/README.md [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+# Timer Trigger
+This trigger provides your flogo application the ability to get data from a CSV file after a particular interval for an action 
+
+## Installation
+
+```bash
+flogo install github.com/skothari-tibco/csvtrigger
+```
+
+## Metadata
```
Comment (fm-tibco) [inline trigger/testTimer/README.md]: Remove this metadata entry, follow something closer to https://github.com/project-flogo/contrib/tree/master/trigger/rest . README.md

File: trigger/testTimer/README.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  }
+}
+```
+### Details
+
+#### Settings:
+| Setting  | Required | Description |
+|:---------|:---------|:------------|
+| control  | false    | Enable control of trigger
+| port     | false    | Port number on which control api is called (/control/resume, /control/pause, /control/restart)
```
Comment (fm-tibco) [inline trigger/testTimer/README.md]: Add a section in the README.md that describes the control aspect.  I wouldn't list those here.  In that section also explain specifying handlerId in order to control a specific handler.

File: trigger/testTimer/README.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+    ]
+  }
+}
+```
+### Details
+
+#### Settings:
+| Setting  | Required | Description |
+|:---------|:---------|:------------|
+| control  | false    | Enable control of trigger
```
Comment (fm-tibco) [inline trigger/testTimer/README.md]: I would remove this, the control should always be enabled, You could consider an autostart flag instead

File: trigger/testTimer/README.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
+<!--
+title: CSV Timer
```
Comment (fm-tibco) [inline trigger/testTimer/README.md]: Fix this and rename the this trigger.  Maybe change it to "streamTester"

File: trigger/testTimer/README.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+|:---------|:---------|:------------|
+| control  | false    | Enable control of trigger
+| port     | false    | Port number on which control api is called (/control/resume, /control/pause, /control/restart)
+
+#### Handler Settings:
+| Setting        | Required | Description |
+|:---------------|:---------|:------------|
+| filePath       | true     | Path to a CSV file
+| repeatInterval | false    | the repeat interval (1, 200 etc in millisecond), doesn't repeat if not specified
+| id             | true     | Id of Handler
```
Comment (fm-tibco) [inline trigger/testTimer/README.md]: There is already a handler id as part of the handler object.  You don't need this.

File: trigger/testTimer/handler.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+		msg := ""
+		if settings == nil {
+			go func() { _ = <-ch }()
+			msg = "The triggers are resumed"
+		} else {
+			go func() { _ = <-settings.Ch }()
+			msg = "The trigger with id "+ settings.Id +" is resumed"
+		}
+		time.Sleep(1 * time.Millisecond)
+		w.WriteHeader(200)
```
Comment (fm-tibco) [inline trigger/testTimer/handler.go]: consider using http.StatusOK

File: trigger/testTimer/timer.go [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+
+	if t.settings.Control {
+		router := httprouter.New()
+
+		//Register For All.
+		router.POST("/control/resume/", resumeHandler(t.ch, nil))
+		router.POST("/control/pause/", pauseHandler(t.ch, nil))
+		router.POST("/control/start/", startHandler(t.ch, nil))
+		router.POST("/control/stop/", stopHandler(t.ch, nil))
+		for _, handler := range ctx.GetHandlers() {
```
Comment (fm-tibco) [inline trigger/testTimer/timer.go]: Maybe change "control" to "tester"
		
router.POST("/tester/resume", resumeHandler(t.ch, nil))
router.POST("/tester/pause", pauseHandler(t.ch, nil))
router.POST("/tester/start", startHandler(t.ch, nil))
router.POST("/tester/stop", stopHandler(t.ch, nil))

Maybe use the following pattern and have the handler switch on id.  You can get the value of the 'id' from httprouter.Params 

router.POST("/tester/resume/:id", resumeHandler(t.ch, nil))
router.POST("/tester/pause/:id", pauseHandler(t.ch, nil))
router.POST("/tester/start/:id", startHandler(t.ch, nil))
router.POST("/tester/stop/":id, stopHandler(t.ch, nil))

File: trigger/testTimer/metadata.go [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+type HandlerSettings struct {
+	Id 			   string `md:"id,required"`
+	StartInterval  string `md:"startDelay"`
+	RepeatInterval string `md:"repeatInterval,required"`
+	FilePath       string `md:"filePath"`
+	Header         bool   `md:"header"`
+	Block          bool   `md:"block,required"`
+	Count          int
```
Comment (fm-tibco) [inline trigger/testTimer/metadata.go]: You are overloading HandlerSettings here.  It should just be actual settings.  I would suggest making another object which you can use to keep track of the chan, data and count.  Maybe you can have a map for ones with id, and a list of all of them to easily start/stop everything.

File: go.mod (hunk 2 above)
Comment (fm-tibco) [reply go.mod]: exclude this go.mod from your PR

File: go.sum [hunk 12]
```diff
@@ -1,10 +1,21 @@
+github.com/davecgh/go-spew v1.1.0 h1:ZDRjVQ15GmhC3fiQ8ni8+OwkZQO4DARzQgrnXU1Liz8=
```
Comment (fm-tibco) [inline go.sum]: Remove this file from the PR

File: trigger/testTimer/README.md (hunk 7 above)
Comment (fm-tibco) [reply trigger/testTimer/README.md]: Actually make it all lowercase "streamtester"
